nncf/quantization/algorithms/hyperparameter_tuner/algorithm.py

```diff
@@ -275,58 +276,101 @@ def apply(self, model: TModel, validation_dataset: Dataset) -> TModel:
+            step_graph = NNCFGraphFactory.create(step_model)
+
+            # If there are no parameters to optimize for the current step, simply execute
+            # this pipeline step on the model.
+            if not step_param_grid:
+                # TODO(andrey-churkin): Think about how it can be avoided.
+                params = apply_combination(self._init_params, best_settings)
+                pipeline = self._pipeline_fn(**params)
+                container = pipeline.get_statistic_points_for_step(step_index, step_model, step_graph)
+                step_statistics = collect_statistics(container, step_model, step_graph, self._calibration_dataset)
+                step_model = pipeline.run_step(step_index, step_statistics, step_model, step_graph)
+                continue
 
-        initial_graph = NNCFGraphFactory.create(model)
+            step_combinations = create_combinations(step_param_grid)
 
-        nncf_logger.info("Start initialization of algorithms")
-        with timer():
-            self._prepare_algorithms(model, initial_graph, combinations)
+            nncf_logger.info(f"Start preparation for {step_index}-th pipeline step")
+            with timer():
+                self._prepare_pipeline_step(step_index, step_model, step_graph, step_combinations, best_settings)
 
-        combination_score_fn = functools.partial(
-            self._calculate_combination_score,
-            initial_model=model,
-            initial_graph=initial_graph,
-            dataset=validation_dataset,
-            subset_indices=subset_indices,
-        )
+            combination_score_fn = functools.partial(
+                self._calculate_combination_score,
+                step_index=step_index,
+                step_model=step_model,
+                step_graph=step_graph,
+                dataset=validation_dataset,
+                subset_indices=subset_indices,
+            )
 
-        nncf_logger.info("Start search best combination of parameters")
-        with timer():
-            best_combination_key = find_best_combination(combinations, combination_score_fn, self._param_grid)
+            nncf_logger.info("Start search best combination of parameters")
+            with timer():
+                step_best_combination_key = find_best_combination(
+                    step_combinations, combination_score_fn, step_param_grid
+                )
+
+            best_settings.update(step_combinations[step_best_combination_key])
+            pipeline = self._pipelines[step_best_combination_key]
+            step_model = pipeline.run_step(
+                step_index, self._step_index_to_statistics[step_index], step_model, step_graph
+            )
 
-        algorithm = self._algorithms[best_combination_key]
-        result_model = algorithm.apply(model, initial_graph, self._statistic_points)
+        # TODO(andrey-churkin): Show final best settings
 
-        return result_model
+        return step_model
```

Comment:
When does the `best_settings` apply to the model? I see that at each step `best_settings` updates with the new combination.

Reply:
The algorithm updates `step_model` iteratively. It starts with `step_model` equal to `model`. After that, the algorithm searches for the best parameters for the current step of the pipeline. Once it finds them, it applies the step (with the found best parameters) to `step_model` to obtain the next `step_model`. So, when the `for` statement is finished, it ends up with the final model obtained using the best settings.

Reply:
Does it mean that we always keep two models in the memory: one with the current `best configuration` and the second with the `current configuration` that would be compared to the `best`?

Reply:
When the algorithm optimizes parameters for the (i + 1)-th step, there is the `step_model_{i}` model for which 0 to i steps were applied with optimal parameters.

For the current configuration of parameters, the algorithm applies the (i + 1)-th step with this configuration to `step_model_{i}` model to obtain `current_step_model_{i+1}` model and calculates the score. After that, `current_step_model_{i+1}` model is destroyed. See `HyperparameterTuner._calculate_combination_score()` method for more details.